clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,216 @@
+  kernel_single_task<class kernel14>(F);
+  // Using the same functor twice should be fine
+}
+
+// kernel13 and kernel14 - expect no errors
+// Pass the same lambda to two kernels
+// Test that passing the same lambda to two kernels does not cause an error
+// because the kernel uses do not interfere with each other or invalidate
+// the stable name in any way
+void callkernel13_14() {
+  kernel13_14func([]() {});
+}
+
+template <typename T>
+static constexpr const char *output1 = __builtin_unique_stable_name(T);
+
+#define MACRO12()                \
+  auto l12 = []() { return 1; }; \
+  constexpr const char *l1_output = __builtin_unique_stable_name(l12); // #USN_l12
+
+int main() {
+
+  // kernel1 - expect no error
+  // Evaluate the builtin in a constant context, then call the kernel with
+  // that same type
+  // Tests that passing the lambda to the unique stable name builtin and
+  // then using the lambda in a way that does not  contribute to the kernel
+  // name does not cause an error because the  stable name is not invalidated
+  // in this situation.
+  auto l1 = []() {};
+  constexpr const char *l1_output = __builtin_unique_stable_name(l1);
+  kernel_single_task<class kernel1>(
+      [=]() { l1(); }); // Used in the kernel, but not the kernel name itself
+
+  // kernel2 - expect error
+  // Call to the builtin on a lambda, then using the lambda for a kernel name
+  // Test that passing the lambda to the USN builtin and then using the same
+  // lambda in the naming of a kernel causes a diagnostic on the kernel use due
+  // to the change in results to the stable name.
+  auto l2 = []() { return 1; };
+  constexpr const char *l2_output = __builtin_unique_stable_name(l2); // #USN_L2
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_L2{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel2>(l2); // Used in the kernel name after builtin
+
+  // kernel7 - expect error
+  // Same as kernel6, except make l6 part of naming kernel
+  // Test that passing a lambda to the unique stable name builtin and then
+  // passing a second lambda to the kernel throws an error because the first
+  // lambda is included in the signature of the second lambda, hence attempting
+  // to change the mangling of the kernel
+  auto l7 = []() { return 1; };
+  auto l8 = [](decltype(l7) *derp = nullptr) { return 2; };
+  constexpr const char *l7_output = __builtin_unique_stable_name(l7); // #USN_l7
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_l7{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@#kernel7call{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel7>(l8); // #kernel7call
```

Comment:
```suggestion
  // expected-note@+1{{in instantiation of function template specialization}}
  kernel_single_task<class kernel7>(l8);
```

Reply:
We don't use `#kernel7call` anywhere else, so this one should be done too.

Reply:
Yes. Thank you for the detailed review!

Reply:
np, but we're still not quite there -- you removed `#kernel6call` in https://github.com/erichkeane/llvm-project/commit/dbada48fe4d774b429ad2957640c82a70d4ef628 (good catch, I hadn't noticed that one needed to be removed) but missed this one for `#kernel7call` that should be replaced with an `@+1` instead.